src/core/exec-invoke.c

```diff
@@ -5267,19 +5277,29 @@ int exec_invoke(
         log_command_line(context, params, "Executing", executable, final_argv);
 
         if (params->exec_fd >= 0) {
-                usec_t t = now(CLOCK_MONOTONIC);
+                uint8_t x = 1;
 
                 /* We have finished with all our initializations. Let's now let the manager know that. From this point
-                 * on, if the manager sees POLLHUP on the exec_fd, then execve() was successful. We send a
-                 * timestamp so that the service manager and users can track the precise moment we handed
-                 * over execution of the service to the kernel. */
+                 * on, if the manager sees POLLHUP on the exec_fd, then execve() was successful. */
 
-                if (write(params->exec_fd, &t, sizeof(t)) < 0) {
+                if (write(params->exec_fd, &x, sizeof(x)) < 0) {
                         *exit_status = EXIT_EXEC;
                         return log_exec_error_errno(context, params, errno, "Failed to enable exec_fd: %m");
                 }
         }
 
+        if (params->handoff_timestamp_fd >= 0) {
+                /* Best effort: get a timestamp now and send it upstream */
+                struct dual_timestamp dt;
+                dual_timestamp_now(&dt);
+
+                (void) send(params->handoff_timestamp_fd,
+                            (const usec_t[2]) { dt.realtime, dt.monotonic }, sizeof(usec_t) * 2,
```

Comment:
Isn't sending a complex struct over a pipe a bit risky? For example, is it possible that different compilation parameters/toolchains/etc could produce different on-the-wire format (which would be a problem if an old executor sent a notification to a new manager due to an upgrade happening at exactly the wrong time)?

Previously just the 64bit usec_t counter was sent, which is simple enough. Any reason for sending the realtime one rather than deriving it later? Is it for the case of clock changes between taking the monotonic timestamp and interpreting it?

Reply:
`usec_t` is the same as `uint64_t`, so there is really no reason why sending two `usec_t`s would be any less safe. This seems completely fine.